File: crates/spec_validate/src/discover.rs
//! Fixture discovery module.
//!
//! Walks specs/fixtures directory and parses filenames according to the
//! naming convention: `<kind>__<expect>__<name>.json`

use std::fmt;
use std::fs;
use std::path::{Path, PathBuf};

/// Fixture kind (domain type category).
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Kind {
    EventEnvelope,
    ModuleManifest,
    SearchDocuments,
    AnalyticsEvents,
}

impl Kind {
    /// Parse kind from filename prefix.
    pub fn from_str(s: &str) -> Option<Kind> {
        match s {
            "event_envelope" => Some(Kind::EventEnvelope),
            "module_manifest" => Some(Kind::ModuleManifest),
            "search_documents" => Some(Kind::SearchDocuments),
            "analytics_events" => Some(Kind::AnalyticsEvents),
            _ => None,
        }
    }

    /// Returns all known kinds.
    pub fn all() -> &'static [Kind] {
        &[
            Kind::EventEnvelope,
            Kind::ModuleManifest,
            Kind::SearchDocuments,
            Kind::AnalyticsEvents,
        ]
    }
}

impl fmt::Display for Kind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Kind::EventEnvelope => write!(f, "event_envelope"),
            Kind::ModuleManifest => write!(f, "module_manifest"),
            Kind::SearchDocuments => write!(f, "search_documents"),
            Kind::AnalyticsEvents => write!(f, "analytics_events"),
        }
    }
}

/// Expected validation outcome.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Expect {
    Valid,
    Invalid,
}

impl Expect {
    /// Parse expectation from string.
    pub fn from_str(s: &str) -> Option<Expect> {
        match s {
            "valid" => Some(Expect::Valid),
            "invalid" => Some(Expect::Invalid),
            _ => None,
        }
    }
}

impl fmt::Display for Expect {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Expect::Valid => write!(f, "valid"),
            Expect::Invalid => write!(f, "invalid"),
        }
    }
}

/// A discovered test case.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Case {
    pub kind: Kind,
    pub expect: Expect,
    pub name: String,
    pub path: PathBuf,
}

impl Case {
    /// Returns a display-friendly identifier.
    pub fn id(&self) -> String {
        format!("{}__{}__{}", self.kind, self.expect, self.name)
    }
}

/// Result of parsing a filename.
#[derive(Debug)]
pub enum ParseResult {
    /// Successfully parsed case.
    Ok(Case),
    /// Not a JSON file.
    NotJson(PathBuf),
    /// JSON file but doesn't match convention.
    NoMatch(PathBuf),
}

/// Parse a filename according to the convention (pure parsing, no filesystem checks).
///
/// Convention: `<kind>__<expect>__<name>.json`
///
/// This function only parses the filename string. Use `parse_file` to also
/// check that the path is an existing file.
pub fn parse_filename(path: &Path) -> ParseResult {
    // Must have .json extension
    let extension = path.extension().and_then(|e| e.to_str());
    if extension != Some("json") {
        return ParseResult::NotJson(path.to_path_buf());
    }

    // Get filename without extension
    let stem = match path.file_stem().and_then(|s| s.to_str()) {
        Some(s) => s,
        None => return ParseResult::NoMatch(path.to_path_buf()),
    };

    // Split by double underscore
    let parts: Vec<&str> = stem.split("__").collect();
    if parts.len() != 3 {
        return ParseResult::NoMatch(path.to_path_buf());
    }

    let kind_str = parts[0];
    let expect_str = parts[1];
    let name = parts[2];

    // Parse kind
    let kind = match Kind::from_str(kind_str) {
        Some(k) => k,
        None => return ParseResult::NoMatch(path.to_path_buf()),
    };

    // Parse expectation
    let expect = match Expect::from_str(expect_str) {
        Some(e) => e,
        None => return ParseResult::NoMatch(path.to_path_buf()),
    };

    // Name must not contain double underscore
    if name.contains("__") {
        return ParseResult::NoMatch(path.to_path_buf());
    }

    // Name must not be empty
    if name.is_empty() {
        return ParseResult::NoMatch(path.to_path_buf());
    }

    ParseResult::Ok(Case {
        kind,
        expect,
        name: name.to_string(),
        path: path.to_path_buf(),
    })
}

/// Parse a file path (checks that it's an existing file first).
pub fn parse_file(path: &Path) -> ParseResult {
    if !path.is_file() {
        return ParseResult::NoMatch(path.to_path_buf());
    }
    parse_filename(path)
}

/// Discovery result containing cases and ignored files.
#[derive(Debug, Default)]
pub struct DiscoveryResult {
    pub cases: Vec<Case>,
    pub ignored: Vec<PathBuf>,
}

/// Discover all fixture cases in the given directory.
///
/// Walks the directory recursively, parses filenames, and returns
/// discovered cases sorted by (kind, expect, name).
pub fn discover(fixtures_dir: &Path) -> std::io::Result<DiscoveryResult> {
    let mut result = DiscoveryResult::default();

    walk_directory(fixtures_dir, &mut result)?;

    // Sort cases by (kind, expect, name)
    result.cases.sort();

    Ok(result)
}

fn walk_directory(dir: &Path, result: &mut DiscoveryResult) -> std::io::Result<()> {
    if !dir.is_dir() {
        return Ok(());
    }

    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();

        if path.is_dir() {
            walk_directory(&path, result)?;
        } else {
            match parse_file(&path) {
                ParseResult::Ok(case) => result.cases.push(case),
                ParseResult::NoMatch(p) => {
                    // Only track .json files that don't match
                    if p.extension().and_then(|e| e.to_str()) == Some("json") {
                        result.ignored.push(p);
                    }
                }
                ParseResult::NotJson(_) => {
                    // Silently ignore non-JSON files (README.md, etc.)
                }
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    #[test]
    fn test_parse_valid_event_envelope() {
        let path = Path::new("specs/fixtures/event_envelope__valid__sample.json");
        match parse_filename(path) {
            ParseResult::Ok(case) => {
                assert_eq!(case.kind, Kind::EventEnvelope);
                assert_eq!(case.expect, Expect::Valid);
                assert_eq!(case.name, "sample");
            }
            _ => panic!("Expected Ok"),
        }
    }

    #[test]
    fn test_parse_invalid_event_envelope() {
        let path = Path::new("specs/fixtures/event_envelope__invalid__missing_field.json");
        match parse_filename(path) {
            ParseResult::Ok(case) => {
                assert_eq!(case.kind, Kind::EventEnvelope);
                assert_eq!(case.expect, Expect::Invalid);
                assert_eq!(case.name, "missing_field");
            }
            _ => panic!("Expected Ok"),
        }
    }

    #[test]
    fn test_parse_module_manifest() {
        let path = Path::new("module_manifest__valid__content_pages.json");
        match parse_filename(path) {
            ParseResult::Ok(case) => {
                assert_eq!(case.kind, Kind::ModuleManifest);
                assert_eq!(case.expect, Expect::Valid);
                assert_eq!(case.name, "content_pages");
            }
            _ => panic!("Expected Ok"),
        }
    }

    #[test]
    fn test_parse_search_documents() {
        let path = Path::new("search_documents__valid__sample.json");
        match parse_filename(path) {
            ParseResult::Ok(case) => {
                assert_eq!(case.kind, Kind::SearchDocuments);
                assert_eq!(case.expect, Expect::Valid);
            }
            _ => panic!("Expected Ok"),
        }
    }

    #[test]
    fn test_parse_analytics_events() {
        let path = Path::new("analytics_events__valid__sample.json");
        match parse_filename(path) {
            ParseResult::Ok(case) => {
                assert_eq!(case.kind, Kind::AnalyticsEvents);
                assert_eq!(case.expect, Expect::Valid);
            }
            _ => panic!("Expected Ok"),
        }
    }

    #[test]
    fn test_parse_wrong_extension() {
        let path = Path::new("event_envelope__valid__sample.txt");
        assert!(matches!(parse_filename(path), ParseResult::NotJson(_)));
    }

    #[test]
    fn test_parse_missing_parts() {
        let path = Path::new("event_envelope__valid.json");
        assert!(matches!(parse_filename(path), ParseResult::NoMatch(_)));
    }

    #[test]
    fn test_parse_unknown_kind() {
        let path = Path::new("unknown_type__valid__sample.json");
        assert!(matches!(parse_filename(path), ParseResult::NoMatch(_)));
    }

    #[test]
    fn test_parse_unknown_expect() {
        let path = Path::new("event_envelope__maybe__sample.json");
        assert!(matches!(parse_filename(path), ParseResult::NoMatch(_)));
    }

    #[test]
    fn test_parse_legacy_filename() {
        // Legacy filenames should not match
        let path = Path::new("valid_event_envelope.json");
        assert!(matches!(parse_filename(path), ParseResult::NoMatch(_)));
    }

    #[test]
    fn test_kind_display() {
        assert_eq!(Kind::EventEnvelope.to_string(), "event_envelope");
        assert_eq!(Kind::ModuleManifest.to_string(), "module_manifest");
        assert_eq!(Kind::SearchDocuments.to_string(), "search_documents");
        assert_eq!(Kind::AnalyticsEvents.to_string(), "analytics_events");
    }

    #[test]
    fn test_expect_display() {
        assert_eq!(Expect::Valid.to_string(), "valid");
        assert_eq!(Expect::Invalid.to_string(), "invalid");
    }

    #[test]
    fn test_case_id() {
        let case = Case {
            kind: Kind::EventEnvelope,
            expect: Expect::Valid,
            name: "sample".to_string(),
            path: PathBuf::from("test.json"),
        };
        assert_eq!(case.id(), "event_envelope__valid__sample");
    }
}

File: packages/platform-core/src/spec-validate/discover.ts
// Fixture discovery for the spec-validate harness.
//
// Port of `crates/spec_validate/src/discover.rs`. Walks a directory and
// parses fixture filenames according to the convention:
//
//   `<kind>__<expect>__<name>.json`
//
// where `kind` is one of (event_envelope | module_manifest |
// search_documents | analytics_events) and `expect` is `valid` | `invalid`.
//
// `name` is freeform but must NOT contain double underscores.

import { readdir, stat } from 'node:fs/promises';
import { join, basename, extname, sep } from 'node:path';

/** Fixture kind (domain type category). */
export type Kind =
  | 'event_envelope'
  | 'module_manifest'
  | 'search_documents'
  | 'analytics_events';

export const ALL_KINDS: ReadonlyArray<Kind> = [
  'event_envelope',
  'module_manifest',
  'search_documents',
  'analytics_events',
] as const;

/** Expected validation outcome encoded in the filename. */
export type Expect = 'valid' | 'invalid';

/** A discovered test case, ready to load + validate. */
export interface Case {
  kind: Kind;
  expect: Expect;
  name: string;
  /** Absolute path to the fixture file. */
  path: string;
}

/** Display-friendly identifier for a case (matches Rust `Case::id`). */
export function caseId(c: Case): string {
  return `${c.kind}__${c.expect}__${c.name}`;
}

export type ParseResult =
  | { tag: 'ok'; case: Case }
  | { tag: 'not_json'; path: string }
  | { tag: 'no_match'; path: string };

function asKind(s: string): Kind | null {
  switch (s) {
    case 'event_envelope':
    case 'module_manifest':
    case 'search_documents':
    case 'analytics_events':
      return s;
    default:
      return null;
  }
}

function asExpect(s: string): Expect | null {
  if (s === 'valid' || s === 'invalid') return s;
  return null;
}

/**
 * Parse a filename according to the convention.
 *
 * Pure string parsing — does NOT touch the filesystem. The returned `path`
 * is whatever was passed in. Use {@link discover} to walk a directory.
 */
export function parseFilename(path: string): ParseResult {
  // file_stem / extension semantics matching Rust's std::path::Path.
  const file = basename(path);
  const ext = extname(file); // includes the leading dot, e.g. ".json"
  if (ext !== '.json') {
    return { tag: 'not_json', path };
  }
  const stem = file.slice(0, file.length - ext.length);
  if (stem.length === 0) {
    return { tag: 'no_match', path };
  }

  const parts = stem.split('__');
  if (parts.length !== 3) {
    return { tag: 'no_match', path };
  }
  // After confirming length === 3, indexed access is safe but TS still
  // sees `string | undefined` under noUncheckedIndexedAccess.
  const kindStr = parts[0]!;
  const expectStr = parts[1]!;
  const name = parts[2]!;

  const kind = asKind(kindStr);
  if (kind === null) return { tag: 'no_match', path };
  const expect = asExpect(expectStr);
  if (expect === null) return { tag: 'no_match', path };
  if (name.length === 0) return { tag: 'no_match', path };
  // After splitting on `__`, no part can contain `__`; guard kept for
  // parity with the Rust check.
  if (name.includes('__')) return { tag: 'no_match', path };

  return { tag: 'ok', case: { kind, expect, name, path } };
}

export interface DiscoveryResult {
  cases: Case[];
  /** `.json` files that didn't match the naming convention. */
  ignored: string[];
}

/**
 * Minimal structural shape of `@atlas/logging`'s Logger that we need —
 * just `debug`/`warn`. Kept structural so this module doesn't pull
 * `@atlas/logging` as a hard dependency; callers thread their
 * `ctx.logger` directly.
 */
export interface DiscoverLogger {
  debug(msg: string, fields?: Record<string, unknown>): void;
  warn(msg: string, fields?: Record<string, unknown>): void;
}

export interface DiscoverOptions {
  /** Optional structured logger for filesystem read failures. */
  logger?: DiscoverLogger;
}

/** Recursively walk `fixturesDir` and return parsed cases sorted stably. */
export async function discover(
  fixturesDir: string,
  options: DiscoverOptions = {},
): Promise<DiscoveryResult> {
  const result: DiscoveryResult = { cases: [], ignored: [] };
  await walk(fixturesDir, result, options.logger);
  result.cases.sort(compareCase);
  return result;
}

async function walk(
  dir: string,
  out: DiscoveryResult,
  logger?: DiscoverLogger,
): Promise<void> {
  let entries: string[];
  try {
    entries = await readdir(dir);
  } catch (err) {
    // Mirror Rust: skip non-dirs without failing the harness, but emit
    // a structured debug record so downstream operators can see why a
    // path returned no cases instead of staring at silence.
    logger?.debug('spec-validate.discover: readdir failed', {
      dir,
      error: (err as Error)?.message ?? String(err),
    });
    return;
  }
  for (const entry of entries) {
    const full = join(dir, entry);
    let s;
    try {
      s = await stat(full);
    } catch (err) {
      logger?.debug('spec-validate.discover: stat failed', {
        path: full,
        error: (err as Error)?.message ?? String(err),
      });
      continue;
    }
    if (s.isDirectory()) {
      await walk(full, out, logger);
      continue;
    }
    const parsed = parseFilename(full);
    switch (parsed.tag) {
      case 'ok':
        out.cases.push(parsed.case);
        break;
      case 'no_match':
        // Only track .json files in the ignored list (matches Rust).
        if (extname(parsed.path) === '.json') out.ignored.push(parsed.path);
        break;
      case 'not_json':
        // Silently ignored.
        break;
    }
  }
}

function compareCase(a: Case, b: Case): number {
  const k = compareKind(a.kind, b.kind);
  if (k !== 0) return k;
  const e = compareExpect(a.expect, b.expect);
  if (e !== 0) return e;
  if (a.name < b.name) return -1;
  if (a.name > b.name) return 1;
  return a.path < b.path ? -1 : a.path > b.path ? 1 : 0;
}

function compareKind(a: Kind, b: Kind): number {
  const ai = ALL_KINDS.indexOf(a);
  const bi = ALL_KINDS.indexOf(b);
  return ai - bi;
}

function compareExpect(a: Expect, b: Expect): number {
  if (a === b) return 0;
  return a === 'valid' ? -1 : 1;
}

// `sep` re-export keeps the import surface minimal — consumers that just
// want to construct fixture paths can import { join } from 'node:path'.
void sep;
